### PR TITLE
fixes for CONFIG_BOOT_TIME_MEASUREMENT

### LINKED ISSUE #19144: arch/x86: CONFIG_BOOT_TIME_MEASUREMENT broken
The problem appears to be an assumption some of the arch-specific benchmark points about the source of timing data.  For example, the datum taken in `crt0.S` assumes the TSC and system timer driver are synchronized and at the same frequency; this is actually not true for most (or all) of our existing x86 platforms.

**To reproduce:**

Build `test/benchmarks/boot_time` for `qemu_x86` and `make run.`  You'll get something like:

```***** Booting Zephyr OS build zephyr-v2.0.0-345-g1aa696bc2be2 *****
starting test - Boot Time Measurement
Boot Result: Clock Frequency: 100 MHz
__start       : 153834280 cycles, 1538342 us
_start->main(): 4141212336 cycles, 4121631054 us
_start->task  : 4141213557 cycles, 4121631066 us
_start->idle  : 4142167642 cycles, 4121640607 us
Boot Time Measurement finished
PASS - main
```

which is obviously bogus.  Similar results on real hardware like an `up_squared`.

### PR DESCRIPTION
**Pull request by ghost**
The boot time measurement sample was giving bogus values on x86: an
assumption was made that the system timer is in sync with the CPU TSC,
which is not the case on most x86 boards.

Boot time measurements are no longer permitted unless the timer source
is the local APIC. To avoid issues of TSC scaling, the startup datum
has been forced to 0, which is in line with the ARM implementation
(which is the only other platform which supports this feature).

Cleanups along the way:

As the datum is now assumed zero, some variables are removed and
calculations simplified. The global variables involved in boot time
measurements are moved to the kernel.h header rather than being
redeclared in every place they are referenced. Since none of the
measurements actually use 64-bit precision, the samples are reduced
to 32-bit quantities.

In addition, this feature has been enabled in long mode.

Fixes: #19144

### CODE REVIEW

File: arch/x86/core/intel64/locore.S [hunk 1]
```diff
@@ -31,8 +31,16 @@ __start:
 1:
 #endif
 
+	/* FIXME: ...resetting TSC here is common to IA32 and Intel64... */
 	/* FIXME: ...disabling the PICs is common to IA32 and Intel64... */
 
+#ifdef CONFIG_BOOT_TIME_MEASUREMENT
+	xorl	%eax, %eax
+	xorl	%edx, %edx
+	movl	$X86_TIME_STAMP_COUNTER_MSR, %ecx
```
Comment (andrewboie) [inline arch/x86/core/intel64/locore.S]: indentation looks off here

File: tests/benchmarks/boot_time/src/main.c [hunk 2]
```diff
@@ -45,37 +35,29 @@ void main(void)
...
 	/* Only print lower 32bit of time result */
 	TC_PRINT("Boot Result: Clock Frequency: %d MHz\n",
 		 freq);
-	TC_PRINT("__start       : %u cycles, %u us\n",
-		 (u32_t)(__start_time_stamp & 0xFFFFFFFFULL),
-		 (u32_t) (_start_us  & 0xFFFFFFFFULL));
 	TC_PRINT("_start->main(): %u cycles, %u us\n",
-		 (u32_t)(s_main_time_stamp & 0xFFFFFFFFULL),
-		 (u32_t)  (main_us  & 0xFFFFFFFFULL));
+		 __main_time_stamp,
```
Comment (andrewboie) [inline tests/benchmarks/boot_time/src/main.c]: here as well

**Review by tyrantlizard77 — COMMENTED** [review]
assuming GH is rendering properly I'm seeing some tabs/spaces issues but otherwise LGTM

File: arch/x86/core/intel64/locore.S (hunk 1 above)
Comment (gnuless) [reply arch/x86/core/intel64/locore.S]: Whoops, I hastily airlifted this from the ia32 version. Fixed the spacing between the mnemonic and operands.

**Review by wentongwu — APPROVED** [review]

File: tests/benchmarks/boot_time/src/main.c (hunk 2 above)
Comment (gnuless) [reply tests/benchmarks/boot_time/src/main.c]: Not sure what you were seeing here -- if it was an artifact of github (I've seen that before) or a genuine ugly -- but I've reformatted with a slightly different style.

File: include/kernel.h [hunk 3]
```diff
@@ -23,6 +23,11 @@
 extern "C" {
 #endif
 
+#ifdef CONFIG_BOOT_TIME_MEASUREMENT
```
Comment (nashif) [inline include/kernel.h]: None of this is public API material, can we put this somewhere else other than kernel.h which is a public API header?

Comment (gnuless) [reply include/kernel.h]: I'm not entirely sure where to put it.  These variables are defined and manipulated from `kernel/`, and used in `tests/` (which probably should be `samples/`, how is this a test?), and are enabled by an option in `subsys/debug/`.  Whoever originally put this in seemed confused about its categorization, as well.

Previously these variables were simply redeclared in all these places, but I don't think going back to that is a good solution.

I'm open to suggestions.

Comment (gnuless) [reply include/kernel.h]: How about `kernel/include/kernel_internal.h` ?

**Review by ioannisg — COMMENTED** [review]
The ARM changes is just the clean-up, right?

File: include/kernel.h (hunk 3 above)
Comment (andrewboie) [reply include/kernel.h]: kernel_internal.h seems a good spot.

**Comment by gnuless** [comment]
> The ARM changes is just the clean-up, right?

Yeah, I removed a variable (since it always had the same value everywhere), so I removed the initialization in the ARM code, too.

File: include/kernel.h (hunk 3 above)
Comment (andrewboie) [reply include/kernel.h]: I'm inclined to merge this as-is, I'm cleaning up how the arch interfaces are specified anyway and will take this under consideration.

**Review by tyrantlizard77 — APPROVED** [review]

**Pull request merged**